components/omega/doc/design/TimeMgr.md

```diff
@@ -0,0 +1,970 @@
+starting with:
+- Gregorian
+- Gregorian with no leap year
+- Idealized 360-day calendar with equal 30-day months
+- No calendar (some test cases only need elapsed time in sec)
+Time for standard calendars will assume Coordinated Universal
+Time (UTC).
+
+### 2.4 Required: Time intervals
+
+The model must be able to compute time intervals, not only to
+represent the model time step itself, but also to be able to
+compute time between events, track periodic events or perform
+time averaging. These intervals can be both absolute time (seconds)
+or represented as calendar intervals (monthly, yearly, number of
+days). Conversion from a calendar interval to absolute interval
+(in seconds) will be needed to compute time-related quantities
+like time averaging.
+
+### 2.5 Required: Events or alarms
+
+The time manager must be able to trigger events that occur
+at specific times or at specified periodic time intervals.
+This requires being able to compare time instants and trigger an
+alarm if a time instant or interval has been reached. It will
+also be useful to track when a previous or next event occurs in a
+periodic sequence.
+
+### 2.6 Required: Long times
+
+To support very long time integrations or potential paleoclimate
+simulations, the model must support years that extend beyond the
+4-digit representation and support negative years to represent
+years Before the Common Era (BCE)
+
+### 2.6 Required: Time string representation
+
+For model output or metadata, the model must be able to
+represent a time in accepted formats. Following ISO standards
+(eg ISO 8601 and prior), these will typically represent time
+in decreasing intervals, eg `YYYYYY-MM-DD HH:MM:SS.SSSSS` though
+variations on this format can be supported to use, for example,
+different separators and different widths for years and fractional
+seconds.
+
+## 3 Algorithmic Formulation
+
+The requirement for no roundoff accumulation implies the need to
+use an integer fraction representation as in the ESMF time
+manager. Similarly, calendar days are best tracked using the
+Julian Day that counts days since a specified start time
+(Noon UTC on 1 Jan, 4713 BCE in the Gregorian calendar - 
+a beautiful Monday by most accounts) with conversion to
+various calendars following the algorithms in:
+
+Fliegel, H. F., and Van Flandern, T. C., 1968, A Machine Algorithm
+for Processing Calendar Dates, Communications of the Association of
Computing Machines, 11, 657.
+
+Hatcher, D.A., 1984. Simple Formulae for Julian Day Numbers and
+Calendar Dates, Quart. J. of R. Astr. Soc., 25, 53-55.
+
+## 4 Design
+
+The time manager design will follow an ESMF-like approach,
+though simplified and not requiring linking to the full
+ESMF library. Previous simplifications were performed for
+WRF by John Michalakes in Fortran and are used in MPAS, E3SM.
+A similar simplification was performed by Phil Jones from the
+ESMF C++ version for the SciDAC CANGA project and will be used
+here.
```

Comment:
@philipwjones just to confirm, this means that the work proposed here for implementing the time manager is in the interface and creating the classes, but that the internal calendar math will be provided by a subset of the ESMF libraries? That sounds great.

Reply:
@mark-petersen  Actually, no. There isn't a way to do that. But I had already done a simplified C++ version for CANGA that I will be using here.

Reply:
Great! The important thing is the underlying library work is already done. Thanks!